### PR TITLE
[core] Fix bug in spilling objects that have empty data field

### LINKED ISSUE #25085: [Core] Infinite retry when spill fails
### What happened + What you expected to happen

When spilling an object fails due to 
```
2022-05-20 14:45:03,120 WARNING worker.py:1416 -- Traceback (most recent call last):
  File "python/ray/_raylet.pyx", line 896, in ray._raylet.spill_objects_handler
  File "python/ray/_raylet.pyx", line 899, in ray._raylet.spill_objects_handler
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/external_storage.py", line 666, in spill_objects
    return _external_storage.spill_objects(object_refs, owner_addresses)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/external_storage.py", line 307, in spill_objects
    return self._write_multiple_objects(f, object_refs, owner_addresses, url)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/external_storage.py", line 137, in _write_multiple_objects
    raise ValueError(error)
ValueError: object ref 1c367dd987a9a540ffffffffffffffffffffffff0300000023000000 does not exist.
An unexpected internal error occurred while the IO worker was spilling objects: object ref 1c367dd987a9a540ffffffffffffffffffffffff0300000023000000 does not exist
```
seems it just keeps retrying.

Might be related to https://github.com/ray-project/ray/issues/17017

https://console.anyscale.com/o/anyscale-internal/projects/prj_JS35jSrv5JFxMqSZmQG7Yif7/clusters/ses_AAwENUbpq7JbYX5ueqMd55p5?command-history-section=command_history

### Versions / Dependencies

master

### Reproduction script

push-based shuffle chaos test

### Issue Severity

_No response_

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Ray sometimes stores errors as the object value in shared memory. These objects have no data since the error is stored in the metadata field. #25085 describes a bug where these objects fail to spill because the IO worker assumes that the data field must be non-empty. This would cause head-of-line blocking for any other objects to spill and cause the whole job to hang. This PR fixes the issue by spilling these objects anyway.

## Related issue number

Closes #25085.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/external_storage.py [hunk 1]
```diff
@@ -128,22 +128,22 @@ def _write_multiple_objects(
...
-                error = f"object ref {ref.hex()} does not exist."
-                # empty data and 1 byte metadata, this object is
-                # marked as failed.
-                if metadata_len == 1:
-                    error += " This is probably since its owner has failed."
+            if buf is None and len(metadata) == 0:
+                error = f"Object {ref.hex()} does not exist."
                 raise ValueError(error)
-            buf_len = len(buf)
+            if buf is None:
```
Comment (rkooo567) [inline python/ray/external_storage.py:134]: What about doing one liner here? 
```
buf = len(buf) if buf else 0
```

File: python/ray/external_storage.py [hunk 2]
```diff
@@ -128,22 +128,22 @@ def _write_multiple_objects(
...
             payload = (
                 address_len.to_bytes(8, byteorder="little")
                 + metadata_len.to_bytes(8, byteorder="little")
                 + buf_len.to_bytes(8, byteorder="little")
                 + owner_address
                 + metadata
-                + memoryview(buf)
             )
+            if buf is not None:
+                payload += memoryview(buf)
```
Comment (rkooo567) [inline python/ray/external_storage.py:146]: payload is `bytes` type and it doesn't have an additional copy when we do `+=` right (unlike string)?

**Review by rkooo567 — APPROVED** [review]
Minor comments

File: python/ray/external_storage.py (hunk 2 above)
Comment (stephanie-wang) [reply python/ray/external_storage.py:146]: Hmm not sure actually, but I'll change this to do it inline with the first allocation!

**Pull request merged**